fastpair/rust/src/bluetooth/windows_ble/adapter.rs

```diff
@@ -0,0 +1,177 @@
+            watcher.SetAllowExtendedAdvertisements(true)?;
+        }
+
+        // `futures::channel::mpsc` is like `std::sync::mpsc` but `impl Stream`.
+        let (sender, receiver) = futures::channel::mpsc::channel(16);
+        let sender = Arc::new(std::sync::Mutex::new(sender));
+
+        // `received_handler` closure holds non-owning channel reference, to
+        // ensure `stopped_handler` can close the channel when
+        // `received_handler` is done.
+        let weak_sender = Arc::downgrade(&sender);
+        let received_handler = TypedEventHandler::new(
+            // Move `weak_sender` into closure.
+            move |watcher: &Option<BluetoothLEAdvertisementWatcher>,
+                  event_args: &Option<BluetoothLEAdvertisementReceivedEventArgs>| {
+                if watcher.is_some() {
+                    if let Some(event_args) = event_args {
+                        if let Some(sender) = weak_sender.upgrade() {
+                            match sender.lock().unwrap().try_send(event_args.clone()) {
+                                Ok(_) => (),
+                                Err(err) => {
+                                    error!("Error while handling Received event: {:?}", err)
+                                }
+                            }
+                        }
+                    }
+                }
+
+                Ok(())
+            },
+        );
+
+        // `stopped_handler` closure owns channel reference, can close channel.
+        let mut sender = Some(sender);
+        let stopped_handler = TypedEventHandler::new(
+            // Move `sender` into closure.
+            move |_watcher,
+                  _event_args: &Option<
+                BluetoothLEAdvertisementWatcherStoppedEventArgs,
+            >| {
+                // Drop `sender`, closing the channel.
+                let _sender = sender.take();
+                println!("Watcher stopped receiving BLE advertisements.");
+                Ok(())
+            },
+        );
+
+        watcher.Received(&received_handler)?;
+        watcher.Stopped(&stopped_handler)?;
+        watcher.Start()?;
+
+        // `receiver` is a `futures::channel::mpsc::Receiver`, which implements
+        // `futures::stream::Stream`. This is essentially an async Iterator.
+        // We apply a FilterMap to map from advertisement packet to a future
+        // returning `BleDevice` and filter out undesired connections. We need a
+        // pinned box to satisfy trait bounds for `Stream`.
+        Ok(Box::pin(receiver.filter_map(move |event_args| {
+            //  Move `watcher` into `FilterMap` closure. This ensures `watcher`
+            // is only dropped when the stream is closed.
+            let _watcher = &watcher;
+
+            // Move `event_args` into async block.
+            async move {
+                match event_args.AdvertisementType().ok()? {
+                    BluetoothLEAdvertisementType::NonConnectableUndirected => None,
```

Comment:
Why are these skipped?

Reply:
i suspect fastpair doesn't need these, but i will probably go through this at some point later and make an enum for the types to be exported back to the caller.